README
This material is subject to the LaTeX Project Public License. See
<http://www.latex-project.org/lppl/> for the details of that license.



### Package information

Package name: spelling  
Summary description: support for spell-checking of LuaTeX documents  
Version: v0.41  
Date: 2013-05-25  
License: [LPPL v1.3c](http://www.latex-project.org/lppl/lppl-1-3c.html)  
Maintenance status: maintained  
Current maintainer: Stephan Hennig, <sh2d@arcor.de>  



### Description

This package supports spell-checking of TeX documents compiled with the
LuaTeX engine.  It can give visual feedback in PDF output similar to
WYSIWYG word processors.  The package relies on an external
spell-checker application that can check a plain text file and output a
list of bad spellings.  The package should work with most
spell-checkers, even dumb, TeX-unaware ones.



### Development

The development repository is currently hosted at
[GitHub](https://github.com/sh2d/spelling/).  Code documentation is in
[LuaDoc](http://keplerproject.github.com/luadoc/) format and can be
generated via

    luadoc -d API *.lua

Bugs and a wish list can be found in the
[issue tracker](https://github.com/sh2d/spelling/issues/).  Patches
welcome!


_Happy TeXing!_



### File list

    CHANGES
    LICENSE
    README
    spelling.sty
    spelling-doc.bad
    spelling-doc.tex
    spelling-doc-lst-lua.tex
    spelling-main.lua
    spelling-recurse.lua
    spelling-stage-1.lua
    spelling-stage-2.lua
    spelling-stage-3.lua
    spelling-stage-4.lua



<!--
%%% Local Variables: 
%%% coding: utf-8
%%% mode: markdown
%%% End: 
-->


CHANGES
This material is subject to the LaTeX Project Public License. See
<http://www.latex-project.org/lppl/> for the details of that license.


### v0.41 (2013-05-25)

Fixes:

* Fixed compatibility issue with LuaTeX 0.70.2 that caused text output
  file written to be emtpy.


### v0.4 (2013-05-23)

New features:

* In addition to lists of bad and good spellings, words can be checked
  against user-defined match rules to determine highlighting status.

Changes:

* Removed means to configure EOL character of text output file.
  Standard Lua EOL character is always used, which is platform
  dependent.

* Improved compatibility with recent LuaTeX versions (v0.74 and newer).

* File `<jobname>.spell.xml` is loaded before file `<jobname>.spell.bad`
  (if both files exist).

Fixes:

* Bad words with surrounding punctuation haven't been highlighted.
  Determining whether a word needs to be highlighted is now done by
  checking against the lists of bad and good spellings words as they
  appear in the document with possible punctuation as well as with all
  surrounding punctuation stripped ([GitHub issue 8][ghi#8]).

* Macro `\spellingoutputlinelength` was broken.

* Raising an error when a file cannot be opened for reading causes
  problems when compiling a document for the first time.  Now, only a
  warning is written to the console and log file in that case.

[ghi#8]: https://github.com/sh2d/spelling/issues/8


### v0.3 (2013-02-12)

New:

* [LanguageTool][lt] support: LanguageTool error reports in the XML
  format can be parsed for spelling errors (with the help of the
  [LuaXML][luaxml] package).  LanguageTool is a cross-platform style and
  grammar checker.

[lt]: http://www.languagetool.org/
[luaxml]: http://www.ctan.org/pkg/luaxml

Changes:

* Default file names used by the package have been changed:

        <jobname>.spb   =>   <jobname>.spell.bad
        <jobname>.spg   =>   <jobname>.spell.good
        <jobname>.txt   =>   <jobname>.spell.txt


### v0.2 (2012-12-04)

Fixes:

* File `spelling.lua` could not be found by LaTeX style file
  ([GitHub issue 14][ghi#14]).

[ghi#14]: https://github.com/sh2d/spelling/issues/14


### v0.1 (2012-11-30)

First upload to CTAN.



<!--
%%% Local Variables: 
%%% coding: utf-8
%%% mode: markdown
%%% End: 
-->


spelling.lua
--- spelling.lua
--- Copyright 2012, 2013 Stephan Hennig
--
-- This work may be distributed and/or modified under the conditions of
-- the LaTeX Project Public License, either version 1.3 of this license
-- or (at your option) any later version.  The latest version of this
-- license is in http://www.latex-project.org/lppl.txt
-- and version 1.3 or later is part of all distributions of LaTeX
-- version 2005/12/01 or later.
--
-- See file README for more information.
--


--- Main Lua file.
--
-- @author Stephan Hennig
-- @copyright 2012, 2013 Stephan Hennig
-- @release version 0.3
--


--- Global table of modules.
-- The work of the spelling package can be separated into four
-- stages:<br />
--
-- <dl>
--
-- <dt>Stage 1</dt>
--   <dd><ul>
--     <li>Load bad strings.</li>
--     <li>Load good strings.</li>
--   </ul></dd>
--
-- <dt>Stage 2  (call-back <code>pre_linebreak_filter</code>)</dt>
--   <dd><ul>
--     <li>Tag word strings in node lists before paragraph breaking
--         takes place.</li>
--     <li>Check spelling of strings.</li>
--     <li>Highlight strings with known incorrect spelling in PDF
--         output.</li>
--   </ul></dd>
--
-- <dt>Stage 3  (<code>\AtBeginShipout</code>)</dt>
--   <dd><ul>
--     <li>Store all strings found on built page via tag nodes in text
--         document data structure.</li>
--   </ul></dd>
--
-- <dt>Stage 4  (call-back <code>stop_run</code>)</dt>
--   <dd><ul>
--     <li>Output text stored in text document data structure to a
--         file.</li>
--   </ul></dd>
--
-- </dl>
--
-- The code of the spelling package is organized in modules reflecting
-- these stages.  References to modules are stored in a table.  Table
-- indices correspond to the stages as shown above.  The table of module
-- references is shared in a global table (`PKG_spelling`) so that
-- public module functions are accessible from within external code.<br
-- />
--
-- <ul>
--   <li><code>spelling-stage-1.lua : stage[1]</code></li>
--   <li><code>spelling-stage-2.lua : stage[2]</code></li>
--   <li><code>spelling-stage-3.lua : stage[3]</code></li>
--   <li><code>spelling-stage-4.lua : stage[4]</code></li>
-- </ul>
--
-- @class table
-- @name stage
stage = {}


--- Table of package-wide resources that are shared among several
--- modules.
--
-- @class table
-- @name res
--
-- @field is_bad  Table.<br />
--
-- This table maps all strings known as bad spellings to the value
-- `true`.
--
-- @field is_good  Table.<br />
--
-- This table maps all strings known as good spellings to the value
-- `true`.
--
-- @field text_document  Table.<br />
--
-- Data structure that stores the text of a document.  The text document
-- data structure stores the text of a document.  The data structure is
-- quite simple.  A text document is an ordered list (an array) of
-- paragraphs.  A paragraph is an ordered list (an array) of words.  A
-- word is a single UTF-8 encoded string.<br />
--
-- During the LuTeX run, node lists are scanned for strings before
-- hyphenation takes place.  The strings found in a node list are stored
-- in the current paragraph.  After finishing scanning a node list, the
-- current paragraph is inserted into the text document.  At the end of
-- the LuaTeX run, all paragraphs of the text document are broken into
-- lines of a fixed length and the lines are written to a file.<br />
--
-- Here's the rationale of this approach:
--
-- <ul>
--
-- <li> It reduces file access <i>during</i> the LuaTeX run by delaying
--   write operations until the end.
--
-- <li> It saves space.  In Lua, strings are internalized.  Since in a
--   document, the same words are used over and over again, relatively
--   few strings are actually stored in memory.
--
-- <li> It allows for pre-processing the text document before writing it
--   to a file.
--
-- </ul>
--
-- @field whatsit_uid  Number.<br />
--
-- Unique ID for marking user-defined whatsit nodes created by this
-- package.  The ID is generated at run-time.  See this <a
-- href="https://github.com/mpg/luatexbase/issues/8">GitHub issue</a>.
--
local res = {

  is_bad,
  is_good,
  text_document,
  whatsit_uid,

}


--- Global package table.
-- This global table provides access to package-wide variables from
-- within other chunks.
--
-- @class table
-- @name PKG_spelling
PKG_spelling = {}


--- Package initialisation.
--
local function __init()
  -- Create resources.
  res.is_bad = {}
  res.is_good = {}
  res.text_document = {}
  res.whatsit_uid = 163
  -- Provide global access to package ressources during module loading.
  PKG_spelling.res = res
  -- Load sub-modules:
  -- * bad and good string loading
  stage[1] = require 'spelling-stage-1'
  -- * node list tagging
  -- * spell-checking
  -- * bad string highlighting
  stage[2] = require 'spelling-stage-2'
  -- * text storage
  stage[3] = require 'spelling-stage-3'
  -- * text output
  stage[4] = require 'spelling-stage-4'
  -- Remove global reference to package ressources.
  PKG_spelling.res = nil
  -- Provide global access to module references.
  PKG_spelling.stage = stage
  -- Enable text storage.
  stage[3].enable_text_storage()
end


-- Initialize package.
__init()


spelling-recurse.lua
--- spelling-recurse.lua
--- Copyright 2012, 2013 Stephan Hennig
--
-- This work may be distributed and/or modified under the conditions of
-- the LaTeX Project Public License, either version 1.3 of this license
-- or (at your option) any later version.  The latest version of this
-- license is in http://www.latex-project.org/lppl.txt
-- and version 1.3 or later is part of all distributions of LaTeX
-- version 2005/12/01 or later.
--
-- See file README for more information.
--


--- Helper module for recursing into a node list.
-- This module provides means to recurse into a node list, calling
-- user-provided call-back functions upon certain events.
--
-- @author Stephan Hennig
-- @copyright 2012, 2013 Stephan Hennig
-- @release version 0.4
--
-- @trick Prevent LuaDoc from looking past here for module description.
--[[ Trick LuaDoc into entering 'module' mode without using that command.
module(...)
--]]


-- Module table.
local M = {}


-- Function short-cuts.
local traverse = node.traverse


-- Short-cuts for constants.
local HLIST = node.id('hlist')
local VLIST = node.id('vlist')


--- Scan a node list and call call-back functions upon certain events.
-- This function scans a node list.  Upon certain events, user-defined
-- call-back functions are called.  Call-back functions have to be
-- provided in a table.  Call-back functions and corresponding events
-- are:
--
-- <dl>
--
-- <dt>`vlist_pre_recurse`</dt> <dd>A vlist is about to be recursed
-- into.  Argument is the vlist node.</dd>
--
-- <dt>`vlist_post_recurse`</dt> <dd>Recursing into a vlist has been
-- finished.  Argument is the vlist node.</dd>
--
-- <dt>`hlist_pre_recurse`</dt> <dd>An hlist is about to be recursed
-- into.  Argument is the hlist node.</dd>
--
-- <dt>`hlist_post_recurse`</dt> <dd>Recursing into a hlist has been
-- finished.  Argument is the hlist node.</dd>
--
-- <dt>`visit`</dt> <dd>A node of type other that `vlist` and `hlist`
-- has been found.  Arguments are the head node of the current node
-- (head node of the current branch) and the current node.</dd>
--
-- </dl>
--
-- If a call-back entry in the table is `nil`, the corresponding event
-- is ignored.
--
-- @param head  Node list.
-- @param cb  Table of call-back functions.
local function recurse_node_list(head, cb)
  -- Make call-back functions local identifiers.
  local cb_vlist_pre_recurse = cb.vlist_pre_recurse
  local cb_vlist_post_recurse = cb.vlist_post_recurse
  local cb_hlist_pre_recurse = cb.hlist_pre_recurse
  local cb_hlist_post_recurse = cb.hlist_post_recurse
  local cb_visit_node = cb.visit_node
  -- Iterate over nodes in current branch.
  for n in traverse(head) do
    local nid = n.id
    -- Test for vlist node.
    if nid == VLIST then
      -- Announce vlist pre-traversal.
      if cb_vlist_pre_recurse then cb_vlist_pre_recurse(n) end
      -- Recurse into 'vlist'.
      recurse_node_list(n.head, cb)
      -- Announce vlist post-traversal.
      if cb_vlist_post_recurse then cb_vlist_post_recurse(n) end
    -- Test for hlist node.
    elseif nid == HLIST then
      -- Announce hlist pre-traversal.
      if cb_hlist_pre_recurse then cb_hlist_pre_recurse(n) end
      -- Recurse into 'hlist'.
      recurse_node_list(n.head, cb)
      -- Announce hlist post-traversal.
      if cb_hlist_post_recurse then cb_hlist_post_recurse(n) end
    -- Other nodes.
    else
      -- Visit node.
      if cb_visit_node then cb_visit_node(head, n) end
    end
  end
end
M.recurse_node_list = recurse_node_list


-- Return module table.
return M


spelling-stage-1.lua
--- spelling-stage-1.lua
--- Copyright 2012, 2013 Stephan Hennig
--
-- This work may be distributed and/or modified under the conditions of
-- the LaTeX Project Public License, either version 1.3 of this license
-- or (at your option) any later version.  The latest version of this
-- license is in http://www.latex-project.org/lppl.txt
-- and version 1.3 or later is part of all distributions of LaTeX
-- version 2005/12/01 or later.
--
-- See file README for more information.
--


--- Handle lists of bad and good strings and match rules.
--
-- @author Stephan Hennig
-- @copyright 2012, 2013 Stephan Hennig
-- @release version 0.4
--
-- @trick Prevent LuaDoc from looking past here for module description.
--[[ Trick LuaDoc into entering 'module' mode without using that command.
module(...)
--]]


-- Module table.
local M = {}


-- Import external modules.
local unicode = require('unicode')
local xml = require('luaxml-mod-xml')


-- Function short-cuts.
local Sfind = string.find

local tabinsert = table.insert

local Ufind = unicode.utf8.find
local Ugmatch = unicode.utf8.gmatch
local Usub = unicode.utf8.sub


-- Declare local variables to store references to resources that are
-- provided by external code.
--
-- Table of known bad strings.
local __is_bad
--
-- Table of known good strings.
local __is_good
--
-- Table of bad rules.
local __rules_bad
--
-- Table of good rules.
local __rules_good


--- Generic function for reading bad or good spellings from a file.
-- All data from the file is read into a string, which is then parsed by
-- the given parse function.
--
-- @param fname  File name.
-- @param parse_string  Custom parse function.
-- @param t  Mapping table bad or good spellings should be added to.
-- @param hint  String for info message.  Either `bad` or `good`.
local function __parse_file(fname, parse_string, t, hint)
  local total_c = 0
  local new_c = 0
  local f, err = io.open(fname, 'r')
  if f then
    local s = f:read('*all')
    f:close()
    total_c, new_c = parse_string(s, t)
  else
    texio.write_nl('package spelling: Warning! ' .. err)
  end
  texio.write_nl('package spelling: Info! ' .. total_c .. '/' .. new_c .. ' total/new ' .. hint .. ' strings read from file \'' .. fname .. '\'.')
end


--- Generic function for parsing a string containing a plain list of
-- strings.  Input format are strings separated by new line or carriage
-- return, i.e., one string per line.  All lines found in the list are
-- mapped to the boolean value `true` in the given table.
--
-- @param s  Input string (a list of strings).
-- @param t  Table that maps strings to the value `true`.
-- @return Number of total and new strings found.
local function __parse_plain_list(s, t)
  local total_c = 0
  local new_c = 0
  -- Iterate line-wise through input string.
  for l in Ugmatch(s, '[^\r\n]+') do
    -- Map string to boolean value `true`.
    if not t[l] then
      t[l] = true
      new_c = new_c + 1
    end
    total_c = total_c + 1
  end
  return total_c, new_c
end


--- Parse a plain list of bad strings read from a file.
-- All strings found (words with known incorrect spelling) are mapped to
-- the boolean value `true` in table `__is_bad`.  The format of the
-- input file is one string per line.
--
-- @param fname  File name.
local function parse_bad_plain_list_file(fname)
  __parse_file(fname, __parse_plain_list, __is_bad, 'bad')
end
M.parse_bad_plain_list_file = parse_bad_plain_list_file


--- Parse a plain list of good strings read from a file.
-- All strings found (words with known correct spelling) are mapped to
-- the boolean value `true` in table `__is_good`.  The format of the
-- input file is one string per line.
--
-- @param fname  File name.
local function parse_good_plain_list_file(fname)
  __parse_file(fname, __parse_plain_list, __is_good, 'good')
end
M.parse_good_plain_list_file = parse_good_plain_list_file


--- Get a custom LanguageTool XML handler.
-- The returned XML handler scans LanguageTool XML data for incorrect
-- spellings.  For every incorrect spelling found, the given call-back
-- function is called with the incorrect spelling string as argument.<br
-- />
--
-- XML data is checked for being created by LanguageTool (via attribute
-- <code>software</code> in tag <code>matches</code>).
--
-- @param cb  Call-back function handling incorrect spellings found in
-- XML data.
-- @return XML handler.
local function __get_XML_handler_LanguageTool(cb)

  -- Some flags for checking validity of XML data.  LanguageTool XML
  -- data must declare as being UTF-8 encoded and advertise as being
  -- created by LanguageTool.
  local is_XML_encoding_UTF_8 = false
  local is_XML_creator_LanguageTool = false
  local is_XML_valid = false

  --- Handler object for parsing LanguageTool XML data.
  -- This table contains call-backs used by LuaXML when parsing XML
  -- data.
  --
  -- @class table
  -- @name XML_handler
  -- @field decl  Handle XML declaration.
  -- @field starttag  Handle all relevant tags.
  -- @field endtag  Not used, but mandatory.
  local XML_handler = {

    decl = function(self, text, attr)
      -- Check XML encoding declaration.
      if attr.encoding == 'UTF-8' then
        is_XML_encoding_UTF_8 = true
        is_XML_valid = is_XML_encoding_UTF_8 and is_XML_creator_LanguageTool
      else
        error('package spelling: Error! XML data not in the UTF-8 encoding.')
      end
    end,

    starttag = function(self, text, attr)
      -- Process <matches> tag.
      if text == 'matches' then
        -- Check XML creator is LanguageTool.
        if attr and attr.software == 'LanguageTool' then
          is_XML_creator_LanguageTool = true
          is_XML_valid = is_XML_encoding_UTF_8 and is_XML_creator_LanguageTool
        end
      -- Check XML data is valid.
      elseif not is_XML_valid then
        error('package spelling: Error! No valid LanguageTool XML data.')
      -- Process <error> tags.
      elseif text == 'error' then
        local ruleid = attr.ruleid
        if ruleid == 'HUNSPELL_RULE'
          or ruleid == 'HUNSPELL_NO_SUGGEST_RULE'
          or ruleid == 'GERMAN_SPELLER_RULE'
          or Ufind(ruleid, '^MORFOLOGIK_RULE_')
        then
          -- Extract misspelled word from context attribute.
          local word = Usub(attr.context, attr.contextoffset + 1, attr.contextoffset + attr.errorlength)
          cb(word)
        end
      end
    end,

    endtag = function(self, text)
    end,

  }

  return XML_handler
end


--- Parse a string containing LanguageTool XML data.
-- All incorrect spellings found in the given XML data are mapped to the
-- boolean value `true` in the given table.
--
-- @param s  String containing XML data.
-- @param t  Table mapping incorrect spellings to a boolean.
-- @return Number of total and new incorrect spellings found.
local function __parse_XML_LanguageTool(s, t)
  local total_c = 0
  local new_c = 0

  -- Create call-back for custom LanguageTool XML handler that stores a
  -- bad word in the given table and does some statistics.
  local cb_incorrect_spelling = function(word)
    if not t[word] then
      t[word] = true
      new_c = new_c + 1
    end
    total_c = total_c + 1
  end

  -- Create custom XML handler.
  local XML_handler_LT = __get_XML_handler_LanguageTool(cb_incorrect_spelling)
  -- Create custom XML parser.
  local x = xml.xmlParser(XML_handler_LT)
  -- Parse XML data.
  x:parse(s)
  return total_c, new_c
end


--- Parse LanguageTool XML data read from a file.
-- All strings found in the file (words with known incorrect spelling)
-- are mapped to the boolean value `true` in table `__is_bad`.
--
-- @param fname  File name.
local function parse_XML_LanguageTool_file(fname)
  __parse_file(fname, __parse_XML_LanguageTool, __is_bad, 'bad')
end
M.parse_XML_LanguageTool_file = parse_XML_LanguageTool_file


--- Parse default sources for bad and good strings.
-- All strings found in default sources for words with known incorrect
-- spelling are mapped to the boolean value `true` in table `__is_bad`.
-- All strings found in default sources for words with known correct
-- spelling are mapped to the boolean value `true` in table `__is_good`.
-- Default sources for bad spellings are files `<jobname>.spell.xml` (a
-- LanguageTool XML file) and `<jobname>.spell.bad` (a plain list file).
-- Default sources for good spellings are file `<jobname>.spell.good` (a
-- plain list file).
local function parse_default_bad_and_good()
  local fname, f
  -- Try to read bad spellings from LanguageTool XML file
  -- '<jobname>.spell.xml'.
  fname = tex.jobname .. '.spell.xml'
  f = io.open(fname, 'r')
  if f then
     f:close()
     parse_XML_LanguageTool_file(fname)
  end
  -- Try to read bad spellings from plain list file
  -- '<jobname>.spell.bad'.
  fname = tex.jobname .. '.spell.bad'
  f = io.open(fname, 'r')
  if f then
     f:close()
     parse_bad_plain_list_file(fname)
  end
  -- Try to read good spellings from plain list file
  -- '<jobname>.spell.good'.
  fname = tex.jobname .. '.spell.good'
  f = io.open(fname, 'r')
  if f then
     f:close()
     parse_good_plain_list_file(fname)
  end
end
M.parse_default_bad_and_good = parse_default_bad_and_good


--- Default bad dictionary look-up match rule.
-- This function looks-up both arguments in the list of bad spellings.
-- It returns `true` if either of the arguments is found in the list of
-- bad spellings, otherwise `false`.
--
-- @param raw  Raw string to check.
-- @param stripped  Same as `raw`, but with stripped surrounding
-- punctuation.
-- @return A boolean value indicating a match.
local function __bad_rule_bad_dictionary_lookup(raw, stripped)
  return __is_bad[stripped] or __is_bad[raw]
end


--- Default good dictionary look-up match rule.
-- This function looks-up both arguments in the list of good spellings.
-- It returns `true` if either of the arguments is found in the list of
-- good spellings, otherwise `false`.
--
-- @param raw  Raw string to check.
-- @param stripped  Same as `raw`, but with stripped surrounding
-- punctuation.
-- @return A boolean value indicating a match.
local function __good_rule_good_dictionary_lookup(raw, stripped)
  return __is_good[stripped] or __is_good[raw]
end


--- Load match rule module.
-- Match rule modules are loaded using `require`.  The module table must
-- follow the following convention: Indentifiers of bad match rules
-- start `bad_rule_`.  Indentifiers of good match rules start
-- `good_rule_`.  Other and non-function identifiers are ignore.
--
-- All match rules found in a module are added to the table of bad and
-- good match rules.  Arguments of a match rule function are a raw
-- string and the same string with stripped surrounding punctuation.
--
-- @param fname  Module file name.
local function read_match_rules(fname)
  local bad_c = 0
  local good_c = 0
  local rules = require(fname)
  for k,v in pairs(rules) do
    if type(v) == 'function' then
      if Sfind(k, '^bad_rule_') then
        tabinsert(__rules_bad, v)
        bad_c = bad_c + 1
      elseif Sfind(k, '^good_rule_') then
        tabinsert(__rules_good, v)
        good_c = good_c + 1
      end
    end
  end
  texio.write_nl('package spelling: Info! ' .. bad_c .. '/' .. good_c .. ' bad/good match rules read from module \'' .. fname .. '\'.')
end
M.read_match_rules = read_match_rules


--- Module initialisation.
--
local function __init()
  -- Get local references to package ressources.
  __rules_bad = PKG_spelling.res.rules_bad
  __rules_good = PKG_spelling.res.rules_good
  -- Add default dictionary look-up match rules.
  tabinsert(__rules_bad, __bad_rule_bad_dictionary_lookup)
  tabinsert(__rules_good, __good_rule_good_dictionary_lookup)
  -- Create emtpy lists of known spellings.
  __is_bad = {}
  __is_good = {}
end


-- Initialize module.
__init()


-- Return module table.
return M


spelling-main.lua
--- spelling-main.lua
--- Copyright 2012, 2013 Stephan Hennig
--
-- This work may be distributed and/or modified under the conditions of
-- the LaTeX Project Public License, either version 1.3 of this license
-- or (at your option) any later version.  The latest version of this
-- license is in http://www.latex-project.org/lppl.txt
-- and version 1.3 or later is part of all distributions of LaTeX
-- version 2005/12/01 or later.
--
-- See file README for more information.
--


--- Main Lua file.
--
-- @author Stephan Hennig
-- @copyright 2012, 2013 Stephan Hennig
-- @release version 0.4
--


-- Module identification.
if luatexbase.provides_module then
  luatexbase.provides_module(
    {
      name        = 'spelling',
      date        = '2013/05/23',
      version     = '0.4',
      description = 'support for spell-checking of LuaTeX documents',
      author      = 'Stephan Hennig',
      licence     = 'LPPL ver. 1.3c',
    }
  )
end


--- Global table of modules.
-- The work of the spelling package can be separated into four
-- stages:<br />
--
-- <dl>
--
-- <dt>Stage 1</dt>
--   <dd><ul>
--     <li>Load bad strings.</li>
--     <li>Load good strings.</li>
--     <li>Handle match rules.</li>
--   </ul></dd>
--
-- <dt>Stage 2  (call-back <code>pre_linebreak_filter</code>)</dt>
--   <dd><ul>
--     <li>Tag word strings in node lists before paragraph breaking
--         takes place.</li>
--     <li>Check spelling of strings.</li>
--     <li>Highlight strings with known incorrect spelling in PDF
--         output.</li>
--   </ul></dd>
--
-- <dt>Stage 3  (<code>\AtBeginShipout</code>)</dt>
--   <dd><ul>
--     <li>Store all strings found on built page via tag nodes in text
--         document data structure.</li>
--   </ul></dd>
--
-- <dt>Stage 4  (call-back <code>stop_run</code>)</dt>
--   <dd><ul>
--     <li>Output text stored in text document data structure to a
--         file.</li>
--   </ul></dd>
--
-- </dl>
--
-- The code of the spelling package is organized in modules reflecting
-- these stages.  References to modules are stored in a table.  Table
-- indices correspond to the stages as shown above.  The table of module
-- references is shared in a global table (`PKG_spelling`) so that
-- public module functions are accessible from within external code.<br
-- />
--
-- <ul>
--   <li><code>spelling-stage-1.lua : stage[1]</code></li>
--   <li><code>spelling-stage-2.lua : stage[2]</code></li>
--   <li><code>spelling-stage-3.lua : stage[3]</code></li>
--   <li><code>spelling-stage-4.lua : stage[4]</code></li>
-- </ul>
--
-- @class table
-- @name stage
stage = {}


--- Table of package-wide resources that are shared among several
--- modules.
--
-- @class table
-- @name res
--
-- @field rules_bad  Table.<br />
--
-- This table contains all bad rules.  Spellings can be matched against
-- these rules.
--
-- @field rules_good  Table.<br />
--
-- This table contains all good match rules.  Spellings can be matched
-- against these rules.
--
-- @field text_document  Table.<br />
--
-- Data structure that stores the text of a document.  The text document
-- data structure stores the text of a document.  The data structure is
-- quite simple.  A text document is an ordered list (an array) of
-- paragraphs.  A paragraph is an ordered list (an array) of words.  A
-- word is a single UTF-8 encoded string.<br />
--
-- During the LuTeX run, node lists are scanned for strings before
-- hyphenation takes place.  The strings found in a node list are stored
-- in the current paragraph.  After finishing scanning a node list, the
-- current paragraph is inserted into the text document.  At the end of
-- the LuaTeX run, all paragraphs of the text document are broken into
-- lines of a fixed length and the lines are written to a file.<br />
--
-- Here's the rationale of this approach:
--
-- <ul>
--
-- <li> It reduces file access <i>during</i> the LuaTeX run by delaying
--   write operations until the end.
--
-- <li> It saves space.  In Lua, strings are internalized.  Since in a
--   document, the same words are used over and over again, relatively
--   few strings are actually stored in memory.
--
-- <li> It allows for pre-processing the text document before writing it
--   to a file.
--
-- </ul>
--
-- @field whatsit_uid  Number.<br />
--
-- Unique ID for marking user-defined whatsit nodes created by this
-- package.  The ID is generated at run-time.  See this <a
-- href="https://github.com/mpg/luatexbase/issues/8">GitHub issue</a>.
--
local res = {

  rules_bad,
  rules_good,
  text_document,
  whatsit_ids,

}


--- Global package table.
-- This global table provides access to package-wide variables from
-- within other chunks.
--
-- @class table
-- @name PKG_spelling
PKG_spelling = {}


--- Determine unique IDs for user-defined whatsit nodes used by this
-- package.  Package luatexbase provides user-defined whatsit node ID
-- allocation since version v0.6 (TL 2013).  For older package versions,
-- we start allocating at an arbitrary hard-coded value of 13**8
-- (ca. 2**30).  Note, for compatibility with LuaTeX 0.70.2, the value
-- must be less than 2^31.
--
-- @return Table mapping names to IDs.
local function __allocate_whatsit_ids()
  local ids = {}
  -- Allocation support present?
  if luatexbase.new_user_whatsit_id then
    ids.start_tag = luatexbase.new_user_whatsit_id('start_tag', 'spelling')
    ids.end_tag = luatexbase.new_user_whatsit_id('end_tag', 'spelling')
  else
    local uid = 13^8
    ids.start_tag = uid + 1
    ids.end_tag = uid + 2
  end
  return ids
end


--- Package initialisation.
--
local function __init()
  -- Create resources.
  res.rules_bad = {}
  res.rules_good = {}
  res.text_document = {}
  res.whatsit_ids = __allocate_whatsit_ids()
  -- Provide global access to package ressources during module loading.
  PKG_spelling.res = res
  -- Load sub-modules:
  -- * bad and good string loading
  -- * match rule handling
  stage[1] = require 'spelling-stage-1'
  -- * node list tagging
  -- * spell-checking
  -- * bad string highlighting
  stage[2] = require 'spelling-stage-2'
  -- * text storage
  stage[3] = require 'spelling-stage-3'
  -- * text output
  stage[4] = require 'spelling-stage-4'
  -- Remove global reference to package ressources.
  PKG_spelling.res = nil
  -- Provide global access to module references.
  PKG_spelling.stage = stage
  -- Enable text storage.
  stage[3].enable_text_storage()
end


-- Initialize package.
__init()
